### PR TITLE
bpo-36346: Prepare for removing the legacy Unicode C API.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Add two compile time options: HAVE_UNICODE_WCHAR_CACHE and
USE_UNICODE_WCHAR_CACHE.

USE_UNICODE_WCHAR_CACHE makes the interpreter not using the
wchar_t cache and the legacy Unicode C API.

HAVE_UNICODE_WCHAR_CACHE removes the wchar_t cache and the
legacy Unicode C API that depends on it.

https://bugs.python.org/issue36346

### CODE REVIEW

File: Modules/_ctypes/callproc.c [hunk 1]
```diff
@@ -1274,22 +1274,28 @@ The handle may be used to locate exported functions in this\n\
...
     HMODULE hMod;
 
     if (!PyArg_ParseTuple(args, "U|O:LoadLibrary", &nameobj, &ignored))
         return NULL;
 
-    name = _PyUnicode_AsUnicode(nameobj);
+#if USE_UNICODE_WCHAR_CACHE
+    const WCHAR *name = _PyUnicode_AsUnicode(nameobj);
+#else /* USE_UNICODE_WCHAR_CACHE */
+    WCHAR *name = PyUnicode_AsWideCharString(nameobj, NULL);
```
Comment (scoder) [inline Modules/_ctypes/callproc.c]: Any reason this isn't `const`?

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -4809,10 +4817,12 @@ static PyMethodDef TestMethods[] = {
     {"getargs_y",               getargs_y,                       METH_VARARGS},
     {"getargs_y_star",          getargs_y_star,                  METH_VARARGS},
     {"getargs_y_hash",          getargs_y_hash,                  METH_VARARGS},
+// #if USE_UNICODE_WCHAR_CACHE
```
Comment (scoder) [inline Modules/_testcapimodule.c]: I assume these are still needed somehow? Are they difficult to remove or just still needed by some tests?

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -900,6 +900,10 @@ typedef struct {
 static void
 path_cleanup(path_t *path)
 {
+#if !USE_UNICODE_WCHAR_CACHE
+    PyMem_Free((wchar_t *)path->wide);
```
Comment (scoder) [inline Modules/posixmodule.c]: Should this maybe clear the field before freeing it? (Thinking of any interaction with tracemalloc or things like that?)

**Review by scoder — COMMENTED** [review]
Wow, that's a lot of code that will eventually be removed here. Nice!

File: Modules/_ctypes/callproc.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_ctypes/callproc.c]: It is passed to `PyMem_Free()`.

File: Modules/_testcapimodule.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_testcapimodule.c]: Remnants of debugging.

File: Modules/posixmodule.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c]: I am not sure this is possible, but I will add a code just for the case.

**Comment by methane** [comment]
@serhiy-storchaka Would you split this pull request?
clinic should be updated before adding DeprecationWarning in PyArg_Parse*.

**Comment by serhiy-storchaka** [comment]
> @serhiy-storchaka Would you split this pull request?
> clinic should be updated before adding DeprecationWarning in PyArg_Parse*.

#21223

File: Include/cpython/unicodeobject.h [hunk 4]
```diff
@@ -74,6 +75,16 @@ Py_UNICODE_FILL(Py_UNICODE *target, Py_UNICODE value, Py_ssize_t length) {
...
+#if HAVE_UNICODE_WCHAR_CACHE
+/* Check if substring matches at given offset.  The offset must be
+   valid, and the substring must not be empty. */
+
+#define Py_UNICODE_MATCH(string, offset, substring) \
+    ((*((string)->wstr + (offset)) == *((substring)->wstr)) && \
+     ((*((string)->wstr + (offset) + (substring)->wstr_length-1) == *((substring)->wstr + (substring)->wstr_length-1))) && \
+     !memcmp((string)->wstr + (offset), (substring)->wstr, (substring)->wstr_length*sizeof(Py_UNICODE)))
+#endif /* HAVE_UNICODE_WCHAR_CACHE */
+
```
Comment (methane) [inline Include/cpython/unicodeobject.h]: Please don't resurrect Py_UNICODE_MATCH.

https://github.com/python/cpython/pull/20878#discussion_r440842197

**Review by blesnatalman — APPROVED** [review]

**Review by blesnatalman — APPROVED** [review]

**Review by blesnatalman — APPROVED** [review]

**Comment by iritkatriel** [comment]
https://bugs.python.org/issue36346 is closed. What is the status of this PR?

**Comment by serhiy-storchaka** [comment]
No longer actual after merging #92537.

**Pull request closed without merging**